nav2_map_server/src/map_server.cpp

```diff
@@ -43,83 +78,162 @@ MapServer::~MapServer()
+  occ_pub_.reset();
+  occ_service_.reset();
+  load_map_service_.reset();
+  msg_.reset();
 
   return nav2_util::CallbackReturn::SUCCESS;
 }
 
 nav2_util::CallbackReturn
-MapServer::on_error(const rclcpp_lifecycle::State &)
+MapServer::on_error(const rclcpp_lifecycle::State & /*state*/)
 {
   RCLCPP_FATAL(get_logger(), "Lifecycle node entered error state");
   return nav2_util::CallbackReturn::SUCCESS;
 }
 
 nav2_util::CallbackReturn
-MapServer::on_shutdown(const rclcpp_lifecycle::State &)
+MapServer::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
 {
   RCLCPP_INFO(get_logger(), "Shutting down");
   return nav2_util::CallbackReturn::SUCCESS;
 }
 
+void MapServer::getMapCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav_msgs::srv::GetMap::Request>/*request*/,
+  std::shared_ptr<nav_msgs::srv::GetMap::Response> response)
+{
+  // if not in ACTIVE state, ignore request
+  if (get_current_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Received GetMap request but not in ACTIVE state, ignoring!");
+    return;
+  }
+  RCLCPP_INFO(get_logger(), "Handling GetMap request");
+  response->map = *msg_;
+}
+
+void MapServer::loadMapCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::LoadMap::Request> request,
+  std::shared_ptr<nav2_msgs::srv::LoadMap::Response> response)
+{
+  // if not in ACTIVE state, ignore request
+  if (get_current_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Received LoadMap request but not in ACTIVE state, ignoring!");
+    return;
+  }
+  RCLCPP_INFO(get_logger(), "Handling LoadMap request");
+  // Load from file
+  if (loadMapResponseFromYaml(request->map_url, response)) {
+    occ_pub_->publish(*msg_);  // publish new map
+  }
+}
+
+bool MapServer::loadMapResponseFromYaml(
+  const std::string & yaml_file,
+  std::shared_ptr<nav2_msgs::srv::LoadMap::Response> response)
+{
+  switch (loadMapFromYaml(yaml_file, *msg_)) {
+    case MAP_DOES_NOT_EXIST:
+      response->result = nav2_msgs::srv::LoadMap::Response::RESULT_MAP_DOES_NOT_EXIST;
+      return false;
+    case INVALID_MAP_METADATA:
+      response->result = nav2_msgs::srv::LoadMap::Response::RESULT_INVALID_MAP_METADATA;
+      return false;
+    case INVALID_MAP_DATA:
+      response->result = nav2_msgs::srv::LoadMap::Response::RESULT_INVALID_MAP_DATA;
+      return false;
+    case LOAD_MAP_SUCCESS:
+      // Correcting msg_ header when it belongs to spiecific node
+      updateMsgHeader();
+
+      response->map = *msg_;
+      response->result = nav2_msgs::srv::LoadMap::Response::RESULT_SUCCESS;
+  }
+
+  return true;
+}
+
+void MapServer::updateMsgHeader()
+{
+  msg_->info.map_load_time = now();
+  msg_->header.frame_id = frame_id_;
+  msg_->header.stamp = now();
+}
+
 }  // namespace nav2_map_server
```

Comment:
Is there a reason to not place the `map_server_main.cpp` contents inside of `map_server.cpp`? 

I'm fine with it either way, it just doesn't look like the map server code has alot of value outside of that executable. 

If you'd like to keep it the way it is, lets reorganize the package just a little bit. Can we make 2 directories in `src/` and `include/` named `map_saver` for the map saver files and `map_server` for the map server files? "saver" and "server" look so similar when just quickly glancing at a filesystem so separating them might be good (if there are multiple server files, probably not worth it if reduced to just 1).

I leave it up to you what you'd like to do.

Reply:
I see this is the common practice in navigation2 is to remain main() function and working classes separately in different files (12 from 15 cpp-files containing main() function). So, I suppose to stick to the same style. Therefore I would like to prefer second option you've mentioned: to split map_server and map_saver code in separate directories in `src/` and `include/`. I'll do it in next patch.

Reply:
Makes sense, OK